sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Authorization/SharedAccessCredential.cs

```diff
@@ -0,0 +1,161 @@
+        /// <param name="signatureResource">The fully-qualified identifier for the resource to which this credential is intended to serve as authorization for.  This is also known as the "token audience" in some contexts.</param>
+        ///
+        public SharedAccessCredential(AzureNamedKeyCredential sourceCredential,
+                                      string signatureResource)
+        {
+            Argument.AssertNotNull(sourceCredential, nameof(sourceCredential));
+            Argument.AssertNotNullOrEmpty(signatureResource, nameof(signatureResource));
+
+            _sourceKeyCredential = sourceCredential;
+            _sourceSasCredential = null;
+
+            var (name, key) = sourceCredential;
+            _sharedAccessSignature = new SharedAccessSignature(signatureResource, name, key);
+        }
+
+        /// <summary>
+        ///   Retrieves the token that represents the shared access signature credential, for
+        ///   use in authorization against an Event Hub.
+        /// </summary>
+        ///
+        /// <param name="requestContext">The details of the authentication request.</param>
+        /// <param name="cancellationToken">The token used to request cancellation of the operation.</param>
+        ///
+        /// <returns>The token representing the shared access signature for this credential.</returns>
+        ///
+        public override AccessToken GetToken(TokenRequestContext requestContext,
+                                             CancellationToken cancellationToken)
+        {
+            var signature = Volatile.Read(ref _sharedAccessSignature);
+
+            // If the signature was derived from a precomputed shared access signature,
+            // it should not be extended.  Bypass expiration checks and generate the
+            // token.
+
+            if (string.IsNullOrEmpty(signature.SharedAccessKey))
+            {
+                // Before forming the token, regenerate the signature if the source
+                // credential has been updated.
+
+                if ((_sourceSasCredential != null) && (!string.Equals(signature.Value, _sourceSasCredential.Signature, StringComparison.Ordinal)))
+                {
+                    signature = new SharedAccessSignature(_sourceSasCredential.Signature);
+                    Volatile.Write(ref _sharedAccessSignature, signature);
+                }
+
+                return new AccessToken(signature.Value, signature.SignatureExpiration);
+            }
+
+            // If the signature was derived from a shared key that has been updated, regenerate
+            // the signature.
+
+            if (_sourceKeyCredential != null)
+            {
+                var (name, key) = _sourceKeyCredential;
+
+                if ((!string.Equals(signature.SharedAccessKeyName, name, StringComparison.Ordinal))
+                    || (!string.Equals(signature.SharedAccessKey, key, StringComparison.Ordinal)))
```

Comment:
Is there a possible timing issue here if the credential is updated after this check?

Reply:
There is, I believe its a benign race and don't see a reasonable way to avoid it.

Worst case, if the source credential is updated after this check, you'll send the outdated credentials and auth will be rejected.  That will cause a retry where we enter this check again, see the updated credentials, and use them.   This would be similar to the path where the auth expires before we need to refresh and we end up in this flow.

More likely, the process of rolling the credentials is gradual and the validity period of the old and new credentials overlap to allow for gentle rolling.  In this case, we'd still send the old credentials but wouldn't experience the auth rejection and would pick up the new credentials the next time refresh was needed or when the auth was no longer valid.